sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -96,43 +224,243 @@ public boolean isInputSortRelAndLimitOnly() {
+            .put(TypeName.FLOAT, "getFloat")
+            .put(TypeName.DOUBLE, "getDouble")
+            .put(TypeName.STRING, "getString")
+            .put(TypeName.DATETIME, "getDateTime")
+            .put(TypeName.BOOLEAN, "getBoolean")
+            .put(TypeName.MAP, "getMap")
+            .put(TypeName.ARRAY, "getArray")
+            .put(TypeName.ROW, "getRow")
+            .build();
+
+    private final Expression input;
+    private final Schema inputSchema;
+
+    private InputGetterImpl(Expression input, Schema inputSchema) {
+      this.input = input;
+      this.inputSchema = inputSchema;
+    }
+
+    @Override
+    public Expression field(BlockBuilder list, int index, Type storageType) {
+      if (index >= inputSchema.getFieldCount() || index < 0) {
+        throw new IllegalArgumentException("Unable to find field #" + index);
+      }
+
+      final Expression expression = list.append("current", input);
+      if (storageType == Object.class) {
+        return Expressions.convert_(
+            Expressions.call(expression, "getValue", Expressions.constant(index)), Object.class);
+      }
+      FieldType fromType = inputSchema.getField(index).getType();
+      String getter = typeGetterMap.get(fromType.getTypeName());
+      if (getter == null) {
+        throw new IllegalArgumentException("Unable to get " + fromType.getTypeName());
+      }
+
+      Expression field = Expressions.call(expression, getter, Expressions.constant(index));
+      if (fromType.getTypeName().isDateType()) {
+        field = Expressions.call(field, "getMillis");
+        if (Arrays.equals(fromType.getMetadata(), CalciteUtils.TIME.getMetadata())) {
+          field = Expressions.convert_(field, int.class);
+        } else if (Arrays.equals(fromType.getMetadata(), CalciteUtils.DATE.getMetadata())) {
+          field =
+              Expressions.convert_(
+                  Expressions.modulo(field, Expressions.constant(MILLIS_PER_DAY)), int.class);
+        } else if (fromType.getMetadata() != null) {
+          throw new IllegalArgumentException(
+              "Unknown DateTime type " + new String(fromType.getMetadata(), UTF_8));
+        }
+      } else if (fromType.getTypeName().isCompositeType()
+          || (fromType.getTypeName().isCollectionType()
+              && fromType.getCollectionElementType().getTypeName().isCompositeType())) {
+        field = Expressions.call(WrappedList.class, "of", field);
+      }
+      return field;
+    }
+  }
+
+  private static final DataContext CONTEXT_INSTANCE = new SlimDataContext();
+
+  private static class SlimDataContext implements DataContext {
+    @Override
+    public SchemaPlus getRootSchema() {
+      return null;
+    }
+
+    @Override
+    public JavaTypeFactory getTypeFactory() {
+      return null;
+    }
+
+    @Override
+    public QueryProvider getQueryProvider() {
+      return null;
+    }
+
+    @Override
+    public Object get(String name) {
```

Comment:
Add a comment about what this is?

Reply:
Added. `/* DataContext.get is used to fetch "global" state inside the generated code */`